src/sacn/source.c

```diff
@@ -595,13 +1232,695 @@ etcpal_error_t sacn_source_reset_networking(sacn_source_t handle, uint16_t unive
  * @return The total number of network interfaces for the universe. If this is greater than netints_size, then only
  * netints_size addresses were written to the netints array. If the source or universe were not found, 0 is returned.
  */
-size_t sacn_source_get_network_interfaces(sacn_source_t handle, uint16_t universe, SacnMcastInterface* netints,
+size_t sacn_source_get_network_interfaces(sacn_source_t handle, uint16_t universe, EtcPalMcastNetintId* netints,
                                           size_t netints_size)
 {
-  ETCPAL_UNUSED_ARG(handle);
-  ETCPAL_UNUSED_ARG(universe);
-  ETCPAL_UNUSED_ARG(netints);
-  ETCPAL_UNUSED_ARG(netints_size);
+  size_t total_num_network_interfaces = 0;
+
+#if SACN_SOURCE_ENABLED
+  if (sacn_lock())
+  {
+    // Look up universe state
+    SacnSourceUniverse* universe_state = NULL;
+    if (lookup_source_state(handle, universe, NULL, &universe_state) == kEtcPalErrOk)
+    {
+      total_num_network_interfaces = universe_state->num_netints;
+
+      // Copy out the netints
+      for (size_t i = 0; netints && (i < netints_size) && (i < total_num_network_interfaces); ++i)
+        netints[i] = universe_state->netints[i];
+    }
+
+    sacn_unlock();
+  }
+#endif
+
+  return total_num_network_interfaces;
+}
+
+bool source_handle_in_use(int handle_val, void* cookie)
+{
+  ETCPAL_UNUSED_ARG(cookie);
+
+  return (handle_val == SACN_SOURCE_INVALID) || (lookup_source_state(handle_val, 0, NULL, NULL) == kEtcPalErrOk);
+}
+
+// Needs lock
+etcpal_error_t start_tick_thread()
+{
+  shutting_down = false;
+  EtcPalThreadParams params = ETCPAL_THREAD_PARAMS_INIT;
+
+  if (etcpal_thread_create(&source_thread_handle, &params, source_thread_function, NULL) != kEtcPalErrOk)
+    return kEtcPalErrSys;
```

Comment:
Why not return the error code from etcpal_thread_create()?

Reply:
The only function that uses this, sacn_source_create, documents the possible return codes that can be returned, and for what reason. I believe this line was my effort to translate internal error codes into a code that fits with the documentation. I do similar error-code-translation elsewhere as well. However I should probably do this translation in the actual API function and just return the code directly here.

Reply:
Most applications won't implement logic based on the error code that's returned beyond checking for Ok vs not Ok, and will only use the differences in the other error codes for logging. So it's not completely critical for the documentation to capture every possible error code that will be returned from an API function, it should just make a "best effort" attempt. In RDMnet, I've written some hand-wavy statements like this in documentation:

https://github.com/ETCLabs/RDMnet/blob/5ba7466a05b98de864a0ff66ecf654ee57125970/src/rdmnet/broker/broker_api.cpp#L51

```
/// @return Other codes translated from system error codes are possible.
```

Reply:
Good to know. I'll probably do the same thing in some places. Do you think there are still times when the error code should be "translated" between internal/external contexts? For instance, if I made a mistake in the source implementation and end up calling an internal function incorrectly, and it returns invalid - does it matter if I proceed to pass that error back to the application even if they fed in valid parameters? Probably not a huge difference, although it would make the logs slightly more clear and useful.

Reply:
Good question. In the situation you've mentioned, I think either option would result in an incorrect error code, because we don't have an error code for "there's a bug in the library".

If you translate the error code, it would return kEtcPalErrSys ("An internal library or system call error occurred"), which is not true.
If you don't, it will return Invalid which is also not true.
What will really happen is that the application user will debug, step down into the library, find the bug and file a bug report. So I don't really think it's a big deal.

Reply:
In that case I'll go through the code and simplify a lot of these translations out, and adjust documentation as needed.

Reply:
That is a great discussion folks!  Good to keep in mind for the future.

Reply:
Resolved in https://github.com/ETCLabs/sACN/pull/32/commits/9546fa80cfb781cdedb46e4109a48d7ec78d2fc4.